carball/analysis/stats/possession/per_possession.py

```diff
@@ -0,0 +1,251 @@
+from collections import Counter
+from typing import List, Dict, Tuple
+
+import pandas as pd
+
+from carball.generated.api.stats.team_stats_pb2 import TeamStats
+from ....analysis.stats.stats import BaseStat
+from ....generated.api import game_pb2
+from ....generated.api.player_pb2 import Player
+from ....generated.api.stats.events_pb2 import Hit
+from ....generated.api.stats.per_possession_stats_pb2 import PerPossessionStats
+from ....generated.api.stats.player_stats_pb2 import PlayerStats
+from ....json_parser.game import Game
+
+
+class PerPossessionStat(BaseStat):
+    def __init__(self):
+        super().__init__()
+        self.initialized = False
+
+        self.player_possessions = []
+        self.team_possessions = []
+
+        self.player_possessions_dict = {}
+        self.team_possessions_dict = {}
+
+    def calculate_player_stat(self, player_stat_map: Dict[str, PlayerStats], game: Game, proto_game: game_pb2.Game,
+                              player_map: Dict[str, Player], data_frame: pd.DataFrame):
+        if not self.initialized:
+            self.initialize(proto_game, player_map, data_frame)
+
+        player_possession_stats_dict: Dict[str, List['PlayerPossession']] = {}
+        for player_id, possessions in self.player_possessions_dict.items():
+            stats = self.get_per_possession_stats(possessions)
+
+            player_stat_map[player_id].per_possession_stats.CopyFrom(stats)
+
+            player_possession_stats_dict[player_id] = stats
+
+        for player_id, player_stats in player_stat_map.items():
+            if player_id in player_possession_stats_dict:
+                player_stats.per_possession_stats.CopyFrom(player_possession_stats_dict[player_id])
+
+    def calculate_team_stat(self, team_stat_list: Dict[int, TeamStats], game: Game, proto_game: game_pb2.Game,
+                            player_map: Dict[str, Player], data_frame: pd.DataFrame):
+        for team_is_orange, possessions in self.team_possessions_dict.items():
+            stats = self.get_per_possession_stats(possessions)
+            team_stat_list[team_is_orange].per_possession_stats.CopyFrom(stats)
+
+    def initialize(self, proto_game: game_pb2.Game, player_map: Dict[str, Player],
```

Comment:
My reasoning was that the thing being called is `calculate_player/team_stat`. And because they're rather short, while the `initialize` is rather long, I'd rather put the two short methods up top where they're easier to find. They then call the `initialize` if needed.